### ISSUE TITLE
BEP: Dynamic Frontend Plugins

### ISSUE THREAD

**Issue by Hyperkid123** [backstage/backstage#22423]
### 🔖 Feature description

The dynamic frontend plugins feature is a way of loading additional frontend plugins at runtime, without the requirement of rebuilding and restarting a running backstage instance.

This system should significantly improve frontend plugin management for backstage instances.

The dynamic plugins leverage the declarative UI system to define what a plugin is and how it should be represented in the browser.

### 🎤 Context

https://github.com/backstage/backstage/tree/master/beps/0002-dynamic-frontend-plugins

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by Hyperkid123** [comment]
A new PR was added with some results of the initial exploration of combining different module federation builds tools in a single runtime. 

Code: https://github.com/scalprum/mf-mixing-experiments
PR: https://github.com/backstage/backstage/pull/22429

**Comment by Rugvip** [comment]
@Hyperkid123, regarding https://github.com/backstage/backstage/pull/22513/commits/4d3cbf7877a9eb100c995a7c9ec3472487a17aee#r1472665454: is there any reason not to maximise module sharing? Basically listing every single external dependency as a shared module? Asking just as a way to explore that extreme.

Selecting the correct version range might be a bit tricky, since using the `package.json` ranges risks that you're testing the plugin towards newer versions of your dependencies, but if a shared module brings in an older one that might not be something that the plugin supports in practice. That could be mitigated by switching out the `package.json` dependency bump strategy, but it might also make sense to go with a completely different solution, like using the versions that's currently installed as the base.

**Comment by Hyperkid123** [comment]
@Rugvip  Sure. If you list all available modules from a package, the module federation is smart enough to share and include in the output only modules that are used in the source code. So there is no problem if only a fraction of modules is used. We use this approach in our projects and it works well. You minimize the overhead.

If you share the entire package (for example MUI), tree shaking is disabled and you bundle **everything** even if it's never used.

There are a few scenarios where this becomes problematic.
1. Massive module libraries like icons where you usually use a fraction of available modules will be just added to the build assets. No way around that.
2. Required version mismatch. If a remote module has a shared config set to a specific version, and no module in the shared scope would match the version requirement, the remote will fall back to its own. This means that you are constantly pulling new assets (which might not be required), which has a performance impact. I think the impact becomes even worse in backstage use case where all plugins are preloaded even though the user never interacts with them.

And because AFAIK backstage plugins do pin their dependencies, the version requirements would rarely be met.

I would say either don't share these types of dependencies at all or just share individual modules and transform the imports at build time.

**Comment by Hyperkid123** [comment]
A quick summary of module federation feature support by individual tested packages in https://github.com/scalprum/mf-mixing-experiments.

## Tested packages
- [Rspack](https://www.rspack.dev/)
- [Webpack native](https://webpack.js.org/)
- [Webpack @module-federation/enhanced](https://github.com/module-federation/enhanced)
- [Vite @originjs/vite-plugin-federation](https://github.com/originjs/vite-plugin-federation)


## Dynamic remotes support

|Tool|Out of the box support|Possible to implement custom support|
|---|---|---|
|Rspack|:white_check_mark:|:white_check_mark:|
|Webpack|:white_check_mark:|:white_check_mark:|
|@module-federation/enhanced|:white_check_mark:|:white_check_mark:|
|@originjs/vite-plugin-federation|:x:|:white_check_mark:|
|@module-federation/vite|:question:|:question:|

## Runtime compatibility

- :white_check_mark: Out of the box compatibility
- :warning: Requires custom implementation 
- :x: Not compatible


|x|Rspack|Webpack|@module-federation/enhanced|@originjs/vite-plugin-federation|
|---|---|---|---|---|
|Rspack|-|:white_check_mark:|:white_check_mark:|:warning:|
|Webpack|:white_check_mark:|-|:white_check_mark:|:warning:|
|@module-federation/enhanced|:white_check_mark:|:white_check_mark:|-|:warning:|
|@originjs/vite-plugin-federation|:warning:|:warning:|:warning:|-|

None of Vite builds are compatible with `@module-federation/runtime` package out of the box. An extra bridge is required. Rspack and Webpack (both the core and the `@module-federation/enhanced` packages) are compatible out of the box.

### Module Scope sharing

- :white_check_mark: Out of the box compatibility
- :warning: Requires runtime specification 
- :x: Not compatible

|x|Rspack|Webpack|@module-federation/enhanced|@originjs/vite-plugin-federation|
|---|---|---|---|---|
|Rspack|-|:white_check_mark:|:white_check_mark:|:warning:|
|Webpack|:warning:|-|:warning:|:warning:|
|@module-federation/enhanced|:white_check_mark:|:white_check_mark:|-|:warning:|
|@originjs/vite-plugin-federation|:warning:|:warning:|:warning:|-|

## Feature parity

The best feature parity is between Rspack and Webpack using `@module-federation/enhanced` tooling.

As mentioned previously, we can make all of the above work with each other. But we have to keep in mind, that runtime dynamic remote support is critical for plugins to be truly dynamic. Currently, that is something the Vite plugin does not support and I was not able to find any plugin that officially supports the feature or a future plan.

We could write a custom plugin, or augment the tested and add a custom dynamic remote support.

**Comment by Rugvip** [comment]
Sorry for the radio silence on this for the last lil bit. Love the research into module federation 👍 🎉 I think it's safe to say that we have a way forward there for now right? Likely sticking to Webpack at least in the short term to keep things simple, and use `@module-federation/enhanced`. Switching to Rspack still seems a lil bit early but def a potential path for the future.

I've now had a proper read through of the BEP and summarized my top prio questions in its [current state](https://github.com/backstage/backstage/blob/7caeb2b5aea8c5b3d9b5e0c36aa57156535f557a/beps/0002-dynamic-frontend-plugins/README.md):

> ## Plugin integration into shell applications

Something I've been thinking about a bit is how we actually manage this config in practice. For example, do you think it's worth building in the ability to essentially dynamically install a plugin + config straight away, or just keep them separate for now?

> ## Plugin registry

It's not actually immediately obvious to me what the purpose of the registry is? Could this be simplified to an interface instead, probably with a baseline implementation that grabs manifest locations or similar from config?

Alternatively I wonder what you think about making this part of the app backend in some form? So that the app backend is generally the source of truth for what additional manifests should be loaded in, maybe even doing some caching/hash validation.

> ## Plugin discovery

Scanning what? isn't the plugin registry already be intended to be the source of truth?

> ## Plugin manifest

At a glance it looks like plugin manifests will be simple to generate, i.e. they're just a build step/artifact, it's not something plugin builders usually need to interact with. Do you agree that that's the case?

> ## Plugin storage

The current proposal covers the case where plugins are deployed and then integrated. What do you think about the case where one wants to install a plugin directly, without requiring a deployment step?

**Comment by Hyperkid123** [comment]
> Likely sticking to Webpack at least in the short term to keep things simple, and use @module-federation/enhanced. Switching to Rspack still seems a lil bit early but def a potential path for the future.

Agreed. RSPack seems promising, but it still has a long way to go to.

> Something I've been thinking about a bit is how we actually manage this config in practice. For example, do you think it's worth building in the ability to essentially dynamically install a plugin + config straight away, or just keep them separate for now?

I see no harm in having this out of the box. From the experience in Janus, making a plugin dynamic was in many cases just a question of triggering the new build pipeline with no changes required in the plugin itself. And because of how the new UI system works in backstage the occasional code tweaks we needed to do will be obsolete.

> Alternatively I wonder what you think about making this part of the app backend in some form? So that the app backend is generally the source of truth for what additional manifests should be loaded in, maybe even doing some caching/hash validation.

This is basically what I had in mind. Because backstage is a full stack app, leveraging the backend is a no-brainer IMO. If the config/registry is then exposed via API or embedded into the main HTML file is probably a matter of preference. The important part is, that the registry/config is not cached by the browser or the API cache is always refreshed on a config change.

> Scanning what? Isn't the plugin registry already intended to be the source of truth?

More or less. The discovery should be the mechanism that triggers whatever it is that is responsible for generating the registry. It can be as sible as having a chokidar watching a directory in which plugins are loaded. I think this will be more relevant in a future where some sort "plugin catalog" exists where users can enable/disable plugins in the UI. Then there is a question about where to actually look for plugins that can be installed. I might be thinking a bit ahead here.

> At a glance it looks like plugin manifests will be simple to generate, i.e. they're just a build step/artifact, it's not something plugin builders usually need to interact with. Do you agree that that's the case?

I don't think they will need to interact at all. Plugin metadata can be generated without the dev input. Again, maybe in the future of some plugin catalog, the manifest might be the right place to add some metadata. Adding an option to extend the manifest with custom properties should not be a big deal if we decide it's needed in the future.

> The current proposal covers the case where plugins are deployed and then integrated. What do you think about the case where one wants to install a plugin directly, without requiring a deployment step?

I now realize I was not clear in the BEP. I am working with the assumption that the frontend plugins installation will work similarly if not the same as the backend dynamic plugins. Both frontend and backend plugins are just JS modules so if we can find a common way to handle the installation I think it would benefit all.

Adding plugins without the extra deployment is the end goal we'd like to achieve.

**Comment by Hyperkid123** [comment]
> Without an extra build I would say is the main goal?

@Rugvip Yes, sorry got the terms mixed up. Deployments are handled by whoever is running the instance.
